packages/shared/src/ContextWidgetPresenter.js

```diff
@@ -0,0 +1,111 @@
+
+export function ContextWidgetPresenter ({ widget, t }) {
+  if (!widget) return widget
+
+  return {
+    ...widget,
+    title: widgetTitleResolver({ widget, t })
+  }
+
+}
+
+export function widgetTitleResolver ({ widget, t }) {
+  let title = widget.title
+  if (title && title.startsWith('widget-')) {
+    title = t(title)
+  } else if (!title) {
+    title = widget.viewGroup?.name || widget.viewUser?.name ||
+            widget.viewPost?.title || widget.viewChat?.name ||
+            widget.customView?.name || ''
+  } else {
+    title = title.split(' ').map(word => word.charAt(0).toUpperCase() + word.slice(1)).join(' ')
+  }
+  return title
+}
+
+export function isValidHomeWidget (widget) {
+  if (widget?.viewChat?.id) return true
+  if (widget?.customView?.id) return true
+  if (widget?.view) return true
+
+  return false
+}
+
+export function wrapItemInWidget (item, type) {
+  return {
+    [type]: item,
+    id: 'fake-id-' + crypto.randomUUID()
+  }
+}
+
+export function isWidgetDroppable ({ widget }) {
+  if (widget.type === 'home') return false
+  if (widget.id?.startsWith('fake-id')) return false
+  return true
+}
+
+export function findHomeView (group) {
+  if (!group?.contextWidgets) {
+    throw new Error('Group has no contextWidgets')
+  }
+  const homeWidget = group.contextWidgets.items.find(w => w.type === 'home')
+  return group.contextWidgets.items.find(w => w.parentId === homeWidget.id)
+}
+
+// export function widgetTypeResolver ({ widget }) {
+  export function widgetTypeInferrer ({ widget }) {
+  switch (true) {
```

Comment:
This could be simplified, couldn't it? I don't see a set of cases I see a chain of precedence, something like that: 

```
return (
  widget?.type ||
  widget?.view ||
  widget?.viewGroup ||
  widget?.viewPost ||
  // ...etc,
  'customView' 
)
```

I woudl find that much more clear in a way. May be missing something, but if the null / boolean state of each of these is the issue perhaps that can be remedied upstream?

Reply:
That works fine for .type and .view but by .viewGroup, it would return the whole group object, instead of a string. Right?

Reply:
Gotcha. To me this would still be more clear and highlight the weird case(s) more which I think is a good thing:

```
return (
  widget?.type ||
  widget?.view ||
  (widget?.viewGroup &&  'viewGroup') ||
  widget?.viewPost ||
  // ...etc, 'customView' 
)
```